File: Python/algorithms/sorting/shell_sort.py
"""  Shell Sort is an advanced version of the insertion sort algorithm that improves
 its efficiency by comparing and sorting elements that are far apart. The idea behind Shell Sort is
  to break the original list into smaller sublists,
 sort these sublists, and gradually reduce the gap between the sublist 
 elements until the list is sorted. """
  
def shellSort(arr):

    n = len(arr)

    gap = n // 2

    while gap > 0:

        for i in range(gap, n):

            temp = arr[i]

            j = i
            while j >= gap and arr[j - gap] > temp:

                arr[j] = arr[j - gap]

                j -= gap

            arr[j] = temp

        gap //= 2

if __name__ == "__main__":
    arr = [12, 34, 54, 2, 3]

    print("Array before sorting:")

    print(arr)

    shellSort(arr)
    

    print("\nArray after sorting:")

    print(arr)

File: CPP/algorithms/sorting/Shell_Sort.cpp
#include <bits/stdc++.h>
using namespace std;

void shellSort(vector<int>& arr) {
    int n = arr.size();
    for (int gap = n / 2; gap > 0; gap /= 2) {
        for (int i = gap; i < n; ++i) {
            int temp = arr[i];
            int j = i;
            while (j >= gap && arr[j - gap] > temp) {
                arr[j] = arr[j - gap];
                j -= gap;
            }
            arr[j] = temp;
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<int> a = {170, 45, 75, 90, 802, 24, 2, 66};

    shellSort(a);

    for (int x : a) cout << x << ' ';
    cout << '\n';
    return 0;
}